sys/net/application_layer/gcoap/fileserver.c

```diff
@@ -0,0 +1,376 @@
+     * an assertion */
+    while ((coap_szx2size(request->szx2) > remaining_length) && (request->szx2 > 0)) {
+        request->szx2--;
+        request->blocknum2 <<= 1;
+    }
+}
+
+static ssize_t gcoap_fileserver_file_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len,
+                                           struct requestdata *request)
+{
+    uint32_t etag;
+    int err = stat_etag(request->namebuf, &etag);
+    if (err < 0) {
+        return gcoap_fileserver_errno_handler(pdu, buf, len, err);
+    }
+
+    if (request->etag_sent && etag == request->etag) {
+        gcoap_resp_init(pdu, buf, len, COAP_CODE_VALID);
+        coap_opt_add_opaque(pdu, COAP_OPT_ETAG, &etag, sizeof(etag));
+        return coap_opt_finish(pdu, COAP_OPT_FINISH_NONE);
+    }
+
+    int fd = vfs_open(request->namebuf, O_RDONLY, 0);
+    if (fd < 0) {
+        return gcoap_fileserver_errno_handler(pdu, buf, len, fd);
+    }
+
+    gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+    coap_opt_add_opaque(pdu, COAP_OPT_ETAG, &etag, sizeof(etag));
+    coap_block_slicer_t slicer;
+    _calc_szx2(pdu, buf, len, request);
+    coap_block_slicer_init(&slicer, request->blocknum2, coap_szx2size(request->szx2));
+    coap_opt_add_block2(pdu, &slicer, true);
+    size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+    err = vfs_lseek(fd, slicer.start, SEEK_SET);
+    if (err < 0) {
+        goto late_err;
+    }
+
+    /* That'd only happen if the buffer is too small for even a 16-byte block,
+     * or if the above calculations were wrong.
+     *
+     * Not using payload_len here as that's needlessly underestimating the
+     * space by CONFIG_GCOAP_RESP_OPTIONS_BUF
+     * */
+    assert(pdu->payload + slicer.end - slicer.start <= buf + len);
+    int read = vfs_read(fd, pdu->payload, slicer.end - slicer.start);
+    if (read < 0) {
+        goto late_err;
+    }
+
+    uint8_t morebuf;
+    int more = vfs_read(fd, &morebuf, 1);
+    if (more < 0) {
+        goto late_err;
+    }
+
+    vfs_close(fd);
+
+    slicer.cur = slicer.end + more;
+    coap_block2_finish(&slicer);
+
+    if (read == 0) {
+        /* Rewind to clear payload marker */
+        read -= 1;
+    }
+
+    return resp_len + read;
+
+late_err:
+    vfs_close(fd);
+    coap_hdr_set_code(pdu->hdr, COAP_CODE_INTERNAL_SERVER_ERROR);
+    return coap_get_total_hdr_len(pdu);
+}
+
+static ssize_t gcoap_fileserver_directory_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len,
+                                                struct requestdata *request,
+                                                gcoap_fileserver_entry_t *resource)
+{
+    vfs_DIR dir;
+    coap_block_slicer_t slicer;
+
+    int err = vfs_opendir(&dir, request->namebuf);
+    if (err != 0) {
+        return gcoap_fileserver_errno_handler(pdu, buf, len, err);
+    }
+    DEBUG("gcoap_fileserver: Serving directory listing\n");
+
+    coap_block2_init(pdu, &slicer);
```

Comment:
I think this call to `coap_block2_init()` is not needed because `coap_block_slicer_init()` overrides the slicer with `request->szx2`, which was set when BLOCK2 was already handled in `gcoap_fileserver_handler()`.